### ISSUE TITLE
Champion "Covariant Return Types"

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#49]
- [ ] Proposal added: [proposals/covariant-returns.md](https://github.com/dotnet/csharplang/blob/master/proposals/covariant-returns.md)
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

**Comment by Joe4evr** [comment]
> This would apply to methods and properties, and be supported in classes and interfaces.

The part about support in interfaces, does this mean something like the following?
```cs
public interface IFoo { }
public interface IBar
{
    IFoo Get();
}
public interface IBaz : IFoo { }
public interface IQux : IBar
{
    override IBaz Get(); //or would it be 'override IBaz IBar.Get();'?
}
```
I guess it's lucky that discussions around Default Interface Implementations have [pushed back a little bit on using `override` as a keyword](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-05-31.md#default-interface-members-overriding-or-implementing) for just that. Even if it wasn't, there probably could've been a way for the two to co-exist, but if you're talking about adding one new thing to interfaces, it would be good to consider the other things that are proposed to be added to interfaces, and make sure that one design choice doesn't preclude any other features entirely.

**Comment by JonHanna** [comment]
I find myself wishing C# supported covariant returns very often, and often with code that isn't particularly "clever".

**Comment by TylerBrinkley** [comment]
Covariant returns would be very useful for typed cloning methods, preventing the need to manually create unique protected cloning methods.

**Comment by MohammadHamdyGhanem** [comment]
Or better:
```
public interface IQux : IBar
{
    this Get(); //or would it be 'override IBaz IBar.Get();'?
}

```
or better:
```
public interface IQux : IBar
{
    [IBaz] Get(); //or would it be 'override IBaz IBar.Get();'?
}


```

**Comment by MohammadHamdyGhanem** [comment]
Why didn't this happen yet?
It seems popular enough.

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

> Why didn't this happen yet?
> It seems popular enough.

Because you haven't written it yet.

Seriously, there are a million "good" ideas, modifying the language is a very expensive process and the team has finite resources.  Things take time.

**Comment by MohammadHamdyGhanem** [comment]
@HaloFour 
> Because you haven't written it yet.

If I can write "it", I would rather write a new S# (Smart Sharp) language, where I do ahatever I want :).

Anyway, may God help the team :)

**Comment by yaakov-h** [comment]
Would this need CLR changes, or compiler changes only?

**Comment by BreyerW** [comment]
According to linked proposal document it is purely compiler magic. It is mentioned that compiler will spit two methods instead of one. One virtual override and another who will call this override and cast result to more specific instance so nothing fancy. Though for that reason alternative mentioned in proposal SEEMS to be better for performance but that will come at the cost of ugliness in code (method duplications).

 I wonder if both couldnt be combined into one proposal? So compiler will produce code same to alternative but source code will look like original proposal.

**Comment by NickRedwood** [comment]
For a long time I didn't realise that the problems I had with many of my inheritance/interface heirarchies nearly always came back to this common problem - the lack of return type covariance to allow overriding methods to narrow their return types. This then leads to duplication of methods and properties to satisfy both the base and the subclass contracts, which is a whole lot of general messyness.

So I am very much in favour of this proposal.

**Comment by YairHalberstadt** [comment]
@gafter
What would that involve? Would a proposal containing C# and target IL for a variety of examples be sufficient?

**Comment by gafter** [comment]
@YairHalberstadt That would help. It would also help to understand how this would affect any tools that consume IL, including compilers for other languages.

**Comment by YairHalberstadt** [comment]
Ok. I will get started on a design proposal.

**Comment by ldematte** [comment]
To save you a little time, fell free to recycle the work I did when I was looking at feasibility.
There is already a mock C# syntax (which is really simple), target IL compiled with ILASM and tested with the C# and VB compilers. And with ILASM/DASM and reflector.
A tool that needed modification for sure was Visual Studio: intellisense got confused, producing strage/non intuitive suggestions, but it worked (no crashes).
Do you have a link at my work/gists?

**Comment by ldematte** [comment]
https://github.com/dotnet/roslyn/issues/357#issuecomment-163286864

https://gist.github.com/ldematte/39b9f3119a92dc058799

Are the relevant parts

**Comment by YairHalberstadt** [comment]
Thanks Idematte. I'll definitely use whatever I can from your work.

**Comment by CyrusNajmabadi** [comment]
> A tool that needed modification for sure was Visual Studio: intellisense got confused, producing strage/non intuitive suggestions, but it worked (no crashes).

Can you be more specific?  What strange/unintuitive suggestions are you referring to?

**Comment by ldematte** [comment]
I have just tried it with VS2017 and it works flawlessly, proposing them as overloads (even it is just the return type changing).
I still have a test project with a compiled dll in my dropobox if you want to try it out.

**Comment by YairHalberstadt** [comment]
That would be great thanks!

On 14 Sep 2018 08:35, "Lorenzo Dematté" <notifications@github.com> wrote:

I have just tried it with VS2017 and it works flawlessly, proposing them as
overloads (even it is just the return type changing).
I still have a test project with a compiled dll in my dropobox if you want
to try it out.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/dotnet/csharplang/issues/49#issuecomment-421259138>,
or mute
the thread

**Comment by ldematte** [comment]
https://www.dropbox.com/sh/hw0sgrae83i93lm/AAC_YXcNxjY3q70JExCoAWfOa?dl=0

There is also a draft proposal I forgot I wrote :)
And a console application with some performance tests too (the stub was designed to work efficiently across multiple subclasses)

**Comment by gafter** [comment]
@ldematte Thank you, your draft proposal is very helpful!

**Comment by YairHalberstadt** [comment]
@gafter, @Idematte, @HaloFour, @sharwell
I have written a relatively detailed exploration of 3 different designs for implementing covariant return types in C#.
I would be interested in hearing what you think, and receiving any feedback you have.
I hope it helpful.
https://github.com/YairHalberstadt/csharplang/blob/master/proposals/covariant-returns-exploration

**Comment by YairHalberstadt** [comment]
Sorry, I have just changed the file name slightly to https://github.com/YairHalberstadt/csharplang/blob/master/proposals/covariant-returns-exploration.md
The old link will give you a 404 error

**Comment by lostmsu** [comment]
I think its funny, that Roslyn itself suffers from the lack of this feature: AST modification methods in it, like `MethodDeclarationSyntax.WithModifiers` could have been virtual.

**Comment by YairHalberstadt** [comment]
Current proposals to implement this require using compiler tricks. That was because till now, the runtime has been considered immutable.

However with the move towards .Net Core, runtime changes have been put back on the table.

Would it be possible to look into if this can be supported at a runtime level?

**Comment by HaloFour** [comment]
@YairHalberstadt

By that you mean having the runtime allow for a method with a more specific return type to be used in an override slot, rather than having the compiler emit bridge methods?

If that were to be considered I think it should be bidirectional and also allow contravariant parameters.

**Comment by YairHalberstadt** [comment]
@HaloFour
Yes, and I agree.

**Comment by Joe4evr** [comment]
While we're at it, might it even be possible to extend "return type covariance"/"parameter type contravariance" to generic arguments when they're in the appropriate position?
```cs
// return type constrained by interface, or shape detection at runtime
public Foo<Base> M()
{
    // currently produces CS0029: Cannot implicitly convert type 'Foo<Derived>' to 'Foo<Base>'
    return new Foo<Derived>();
}
```

**Comment by CyrusNajmabadi** [comment]
@Joe4evr I'd prefer `in/out` variance be specifiable on a class, with appropriate checking that the shape of the type fits those rules.

Note: in the interim, you could always get what you've written by using interfaces.  This also ensures you can separate out the part of shape of your type that you want to be variant, without restricting other parts of your type (say, your private methods) that won't need those rules applied to them.

**Comment by andre-ss6** [comment]
How much more work would supporting contravariant parameters involve?

Based on my limited experience, I think that those have way more limited use-cases than covariant return types while at the same time bringing more issues to the table.

**Comment by gafter** [comment]
> How much more work would supporting contravariant parameters involve?

It would involve extreme breaking changes in common scenarios.  In other words, impossible.

**Comment by andre-ss6** [comment]
Well then I guess we don't need to even consider it? 🤷‍♂️

**Comment by IdeaHunter** [comment]
Im trying to sum up current state of this feature, please correct me if im wrong:
1) Sufficiently good(for poking roslyn with a stick) proposal *is* in place  
2) There *was* a successful attempt had been made with manual IL changes as one proposed in feature
https://github.com/YairHalberstadt/csharplang/blob/master/proposals/covariant-returns-exploration.md
3) There *was* a successfull attempt at making it work with vs 2017
https://github.com/dotnet/csharplang/issues/49#issuecomment-421259138
4) There *is* info which roslyn methods are need to be poked to make this feature work
https://github.com/dotnet/roslyn/issues/357#issuecomment-162603202
5) These *isn't* a PR for the feature

This rises a question: 
Is PR for roslyn with this feature welcomed at this stage of discussion?

**Comment by gafter** [comment]
I don't know about 2-4.  I should also point out that the LDM and compiler team haven't decided to prioritize this for any particular release.  Having said that, a full prototype would give us more information about whether this is worth whatever it might turn out to cost.  So sure, a prototype would be helpful.

**Comment by ViIvanov** [comment]
> > How much more work would supporting contravariant parameters involve?
> 
> It would involve extreme breaking changes in common scenarios. In other words, impossible.

@gafter Could you please explain (just small example), what breaking changes you mean?

**Comment by gafter** [comment]
One example of a breaking change that we would have if we add contravariant parameters:

``` c#
class Base
{
    public virtual void M(string o) { }
    public virtual void M(object o) { }
}
class Derived : Base
{
    public override void M(object o) { } // oops, now overrides both methods
}
```

**Comment by Logerfo** [comment]
I like this because it solves some cases where CRTP would be needed. But this doesn't solve all the cases... First of all: does this feature require CLR changes?
The problem with CRTP is that I cannot have polymorphism without actually knowing the constructed type, which kinda breaks its usefulness. For example:
```c#
class Animal<T> where T: Animal<T>, new()
{
   public T[] Parents { get; private set; }
   public T Breed(T other) => new T
   {
      Parents = { this, other }
   };
}
class Dog : Animal<Dog> { }
class Person
{
   public Animal[] Pets { get; } //not possible
}
```
But what if C# had this `T` type for the constructed type out of the box, without the need for an actual generic enclosing type. Like this:
```c#
class Animal
{
   public TConstructed[] Parents { get; private set; }
   public TConstructed Breed(TConstructed other) => new TConstructed
   {
      Parents = { this, other }
   };
}
class Dog : Animal { }
class Person
{
   public Animal[] Pets { get; } //now possible
}
```
It doesn't matter how `TConstructed` is called (can't think of a non-breaking syntax).
The `new()` constraint is debatable but that's just an example anyway.
Does my suggestion require CLR changes? If both requires or if both doesn't, I think the cost-benefit of my suggestion would fit better IMO, since it could be used for virtual methods as well.

**Comment by HaloFour** [comment]
@Logerfo 

Depends on how covariant return types are implemented.  The CLR has always considered the return type of a method to be a part of its signature and for overloads that only differ by return type.  Covariant return types could build on top of this by using an overload to define the override with the covariant return type and emitting a bridge method as the actual overload:

```cs
public abstract class Foo {
    public abstract object DoStuff();
}

public class Bar : Foo {
    public override string DoStuff() => "stuff";
}

// compiler emits
public class Bar : Foo {
    public new string DoStuff() => "stuff";

    override object Foo.DoStuff() => this.DoStuff();
}
```

I used an explicit implementation-like syntax above as C# doesn't currently support a syntax that would enable this.  But such syntax is fitting as the CLR also allows for explicit overrides of virtual methods (complete with the ability to change the accessibility and the name of the overridden method).

As for CRTP I think the question is how it would be implemented.  Auto-generating covariant returning methods seems easy enough, but I don't think it would work well without generics being involved, and that would either require CLR support or some janky C# voodoo which would probably be difficult to prevent from leaking.

I don't think that covariant returns are limited to the CRTP case so it makes sense to enable them as a standalone language feature, especially since they also don't require any CLR changes or even really any changes to the C# grammar.

**Comment by Logerfo** [comment]
@HaloFour I see. I opened another issue (#2839) to split up the suggestions then.

**Comment by gafter** [comment]
I have created https://github.com/dotnet/csharplang/issues/2844 as a draft specification for this feature.

**Comment by gafter** [comment]
Just a note: the LDM is proceeding with this feature with the premise that we can do it with runtime support rather than through the generation of bridge methods.  Our current exploration will help us understand how feasible this approach really is.

**Comment by HaloFour** [comment]
@gafter 

By doing it through runtime support you mean having the CLR allow for the `.override` clause to be allowed when the method has a signature that only differs by return type as long as the return type has identity or implicit reference conversion to the return type of the overridden method?

```il
.class public auto ansi beforefieldinit A extends [mscorlib]System.Object
{
    .method public hidebysig newslot virtual instance object M () cil managed 
    {
        // yadda yadda yadda
    }
}

.class public auto ansi beforefieldinit B extends A
{
    .method public hidebysig virtual instance string M () cil managed 
    {
        .override method instance object A::M()
        // yadda yadda yadda
    }
}
```

**Comment by gafter** [comment]
@HaloFour That is the most likely approach, and something corresponding for interfaces too.

**Comment by gafter** [comment]
Prioritization of runtime support for covariant return types is being tracked internally at https://devdiv.visualstudio.com/DevDiv/_queries/edit/1009909

**Comment by mikernet** [comment]
Please also extend this feature so it doesn't have to be strictly covariant in the currentlt used sense but rather also allows value type return values to implement methods that expect an `object` return type, i.e:

https://github.com/dotnet/csharplang/issues/296

I believe the strict definition of "covariant" currently only applies to reference types, right?

**Comment by gafter** [comment]
FYI, the draft specification is at https://github.com/dotnet/csharplang/blob/master/proposals/covariant-returns.md

**Comment by kinchungwong** [comment]
My two cents.

Repurpose the **`as` keyword** in the method return type specification.

For example:
```C#
public abstract class Foo {
    public abstract object DoStuff();
}

public class ProposedBar : Foo {
    public override string as object DoStuff() => "stuff"; // covariantly overrides Foo.DoStuff()
}
```

The presence of the **`as` keyword** can trigger some compiler magic that will help the compiler understand which interface method or base class abstract method it is supposed to be implementing covariantly.

The use of **`as` keyword** can be made optional. It is only required if the compiler do not have enough information to figure out automatically.

**Comment by gafter** [comment]
@kinchungwong The compiler doesn't need any "help" to understand which method is being overridden. The return type was never part of that calculation.

**Comment by TobiasBengtsson** [comment]
Are there any plans to make it simpler to write functions that returns `typeof(this)`? It seems like with covariant return types, just some syntactic sugar needs to be added. This is one of the most common scenarios for covariant returns I have encountered. E.g.

```
abstract class FluentBuilder
{
    public virtual thistype DoSomething() // new keyword - will return FluentBuilder in this class, but subclass type in subclasses
    {
        // ...
        return this; // Can only be allowed to "return this" for it to work(?)
    }
}
class SomeFluentBuilder : FluentBuilder { } // will not need to override DoSomething to make it return SomeFluentBuilder
```

instead of (C#9)

```
abstract class FluentBuilder
{
    public virtual FluentBuilder DoSomething()
    {
        // ...
        return this;
    }
}
class SomeFluentBuilder : FluentBuilder
{
    public override SomeFluentBuilder DoSomething()
    {
        base.DoSomething();
        return this;
    }
}
```

**Comment by yaakov-h** [comment]
There are other proposals that cover that.

**Comment by J0rgeSerran0** [comment]
One question about Covariant Return Types
I am a little confused with it

Is this feature ready to be used in the last version of NET 5 (5.0.100-rc.1.20452.10 version) or is in progress yet?

**Comment by amis92** [comment]
> One question about Covariant Return Types
> I am a little confused with it
> 
> Is this feature ready to be used in the last version of NET 5 (5.0.100-rc.1.20452.10 version) or is in progress yet?

It's implemented but RC1 contains a bug with using derived type in abstract causing a TypeLoadException. Hit me. RC2 has it fixed.

**Comment by J0rgeSerran0** [comment]
I was writing this because my Covariant Return Types code was not working as expected and I was confused about it
Thanks a lot for the details about the issue with it @amis92

**Comment by J0rgeSerran0** [comment]
I confirm you that, with the .NET 5 RC2 published yesterday, Covariant Return Types is working correctly now

**Comment by Daniel15** [comment]
It sounds like this is now available in C# 9.0 (ref https://devblogs.microsoft.com/dotnet/c-9-0-on-the-record/#covariant-returns) so this issue can probably be closed?

**Comment by erichiller** [comment]
The spec leaves open the question of whether `override` would be supported for interface implementators to have covariant return types, has a decision been made on this?

example, the spec mentions
```cs
interface I1 { I1 M(); }
interface I2 { I2 M(); }
interface I3: I1, I2 { override I3 M(); }
```

But currently `override` is not allowed on interface properties as far as I can tell.

**Comment by TahirAhmadov** [comment]
It's a bit unclear from this thread and other notes. Is it done using "CLR hack" of just allowing covariant return types, or the compiler double-method emittance?

**Comment by HaloFour** [comment]
@TahirAhmadov 

The CLR was modified to allow the overriding method to have a covariant return type.  I don't see why that would be considered a "hack", though.

**Comment by TylerBrinkley** [comment]
Perhaps they thought of the solution as a "hack" since it required users to target a new CLR as opposed to being backwards compatible with older CLR's like that of the .NET Framework.

**Comment by TahirAhmadov** [comment]
> @TahirAhmadov
> 
> The CLR was modified to allow the overriding method to have a covariant return type. I don't see why that would be considered a "hack", though.

You'll laugh but that's how it was characterized by a member of the language team about 10 years ago when I first brought it up, before GitHub.

**Comment by andre-ss6** [comment]
> > @TahirAhmadov
> > The CLR was modified to allow the overriding method to have a covariant return type. I don't see why that would be considered a "hack", though.
> 
> You'll laugh but that's how it was characterized by a member of the language team about 10 years ago when I first brought it up, before GitHub.

Yea, things changed a lot in that time. I understand that the main reason they think differently now is the fact that changing the CLR no longer introduces the risk of automatically breaking every .NET application on your PC.

**Comment by TheConstructor** [comment]
> The [spec](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/covariant-returns) leaves open the question of whether `override` would be supported for interface implementators to have covariant return types, has a decision been made on this?
> 
> example, the spec mentions
> 
> ```cs
> interface I1 { I1 M(); }
> interface I2 { I2 M(); }
> interface I3: I1, I2 { override I3 M(); }
> ```
> 
> But currently `override` is not allowed on interface properties as far as I can tell.

One thing I would love, would be to write `GetEnumerator()` once to implement `IEnumerable` and `IEnumerable<T>`. Would `override`-support in interfaces be required for this?

**Comment by IanKemp** [comment]
This didn't age well.

**Comment by fandrei** [comment]
@Daniel15 
> It sounds like this is now available in C# 9.0 (ref https://devblogs.microsoft.com/dotnet/c-9-0-on-the-record/#covariant-returns) so this issue can probably be closed?

In any case, the part of the proposal about covariance in the interfaces was not implemented. Looks like it just slipped through the cracks.

**Comment by CyrusNajmabadi** [comment]
This did not slip through.  The design is explicitly about overrides: https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/covariant-returns.md

**Comment by fandrei** [comment]
@CyrusNajmabadi also, see here - the conclusion in the end. https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-01-08.md
You decided that this feature is not valuable enough. Apparently, other people on your team had different opinion. However, I didn't find any follow ups. Looks like it did slip through the cracks.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi also, see here - the conclusion in the end. https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-01-08.md
> You decided that this feature is not valuable enough. Apparently, other people on your team had different opinion. However, I didn't find any follow ups. Looks like it did slip through the cracks.

"We like it in principle" doesn't mean "we are doing it". That's the exact point I'm making. Actual value trumps principles every time.

**Comment by fandrei** [comment]
@CyrusNajmabadi I think that it does mean that it's worth further consideration.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I think that it does mean that it's worth further consideration.

There are tons of things we like in principle. The point is that we invest in the thumbs we think provide major value.

**Comment by CyrusNajmabadi** [comment]
> Well it's fallen through the cracks in the sense there is absolutely no movement on this issue for three years

Correct. We invest in the things we think will provide high value for the ecosystem. 

> I tagged a couple of the designers a few months ago, but apparently I'm not even worthy of a response.

I don't know what you want. I'm one of the designers.  I'm responding. I'm being as explicit as possible about how this works and how we assess things. You seem to not be on with this.  But there's nothing I can do about that. 

For this to be done, a highly compelling case needs to be made to DEINVEST in current investments and move to this. Every lang and runtime designer can look at that case and decide if the are moved by it. If so, it can be done.  If not... Well that's life. 

Fwiw, I'm in the design team and I have more failures getting something done than successes.  That's fundamentally just how things work when you have THOUSANDS of possible choices and you have to pick a tiny handful each year.

**Comment by jaredpar** [comment]
> Well it's fallen through the cracks in the sense there is absolutely no movement on this issue for three years, 

Fallen through the cracks generally means there was an intent to do an item but that did not happen. That is not the case here. There has not ever been an explicit plan for this to work in the ecosystem. Also in the last three years this has not once come up as a substantial ecosystem blocker. 

> Well it's fallen through the cracks in the sense there is absolutely no movement on this issue for three years, 

There are lots of issues for which there is no movement.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi well, apparently the `vals is [1, 2, 3]` pattern is something that all the developers desperately needed.

Correct. I would absolutely invest in patterns more than this.

**Comment by CyrusNajmabadi** [comment]
I don't know if any easy ways to do what list patterns accomplish.

**Comment by fandrei** [comment]
@CyrusNajmabadi 
> I don't know of any easy ways to do what list patterns accomplish.

For this particular one written in the docs, you can just write `vals.SequenceEqual()`
For more complex ones - are there any _real_ examples where they are useful?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi 
> >  I'm one of the designers. I'm responding.
> 
> I didn't know that, sorry. It's a shame there isn't some sort of flair system for usernames.

Isn't there?  I should show up as a member of csharplang 

> 
> >  I'm being as explicit as possible about how this works and how we assess things. You seem to not be onkwith this.
> 
> No, that wasn't it, I was merely commenting on the complete lack of movement *up until* today for at least three years. It clearly hasn't been a priority in the past

Correct. It has not been a priority. The same with 99% of the suggestions here. 

. I was saying I poked it to get some attention four months ago but no-one seemed to care at the time.

Correct. People can poke things, but that doesn't change people caring. There are thousands of issues, with tons of pokes. There are limited resources, and and tons to do. Realistically, the vast majority of time that the vast majority is people want *will not get done* or even looked at. This is the reality of the situation 

> 
> > Correct. We invest in the things we think will provide high value for the ecosystem.
> 
> > For this to be done, a highly compelling case needs to be made to DEINVEST in current investments and move to this. Every lang and runtime designer can look at that case and decide if the are moved by it. If so, it can be done. If not... Well that's life.
> 
> > Fwiw, I'm in the design team and I have more failures getting something done than successes. That's fundamentally just how things work when you have THOUSANDS of possible choices and you have to pick a tiny handful each year.
> 
> I understand, and I'm not going to get in a tussle about decisions that have been made in the past. But it's been mentioned on various threads that upvoting is supposed to be a good indication that a proposal will be taken forward, but this doesn't seem to be happening. It also doesn't seem as complex to implement as some of the other things already done (eg DIM, or the first part of this Issue).

Dim had major backing from large teams. Add did covariant returns. 

Interface impl might be super important to you, but it just isn't to others. That's life :-)

> 
> I get the feeling part of the problem is this Issue is marked as "Implemented Needs ECMA Spec" even though it's not fully implemented. So it tends to get ignored as it looks like it's just a holdover from the first part of the implementation. Perhaps the label should be changed.

It is implemented, and we are waiting on the ecma specification party to finally move to the completed state. 

You are saying it is not fully implemented, however the view of both the language and runtime is that we implemented the parts we wanted to implement.  

Perhaps that will grow in the future. But right now, the part you want is simply not something we agree looking at. If that changes, someone will def can take out as a honey priority, based on what we think the ecosystem and partners need :-)

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi 
> > I don't know of any easy ways to do what list patterns accomplish.
> 
> For this particular one written in the docs, you can just write `vals.SequenceEqual()`

That won't work for any of the normal cases I use patterns for.  Like non allocated matching of a complex hierarchy, while pulling out the values of interest and placing into variables. 

> For more complex ones - are there any _real_ examples where they are useful?

I have no way to answer that.  I have no way to determine what you find useful.

**Comment by fandrei** [comment]
@CyrusNajmabadi 
> I have no way to answer that. I have no way to determine what you find useful.

Any example that you find useful would work.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi 
> > I have no way to answer that. I have no way to determine what you find useful.
> 
> Any example that you find useful would work.

`X is [p1, .., p2]`

Is a good example for me.

**Comment by fandrei** [comment]
@CyrusNajmabadi do you mean
```
		var p1 = X[0];
		var p2 = X[^1];
```
?

**Comment by CyrusNajmabadi** [comment]
No. I just generally mean that I'm testing multiple patterns across the entire sequence of input. 

Note: I don't think we go far enough here. I wish that the would work over arbitrary IEnumerable stand streams as well. Those that don't have indexers at all. 

We have a proposal and accepted design for that I think. But we do need to cost/impl if we want to move forward on that.

**Comment by CyrusNajmabadi** [comment]
In your case you are extracting values. I'm taking about arbitrary pattern matching of values in a sequence, just as I can rest arbitrary patterns down an object hierarchy. 

List patterns extend that matching beyond a single value to a collection of values. 

Just as I can use expressions to build up arbitrarily complex patterns (including collections), I want the same ability to break apart arbitrary graphs (including collections) with patterns.

**Comment by jaredpar** [comment]
> Any example that you find useful would work.

How about examples that _customers_ find useful

https://github.com/search?q=%2F+is+%5C%5B%2F+language%3Ac%23&type=code

This regex is imprecise so it will say match comments.. But it also matches many real samples of customers finding value in `is [ ... ]` patterns.

**Comment by fandrei** [comment]
@jaredpar it also matches a lot of the code that constructs strings, and the code that could be easily rewritten in a different way, and the code that is not even C#, etc.
I wanted to ask if you have a way to know how much different features are actually used, using a real C# parser, but it looks like you don't.
Anyway, I don't remember many people asking about list pattern matching in C# on StackOverflow.

**Comment by fandrei** [comment]
@CyrusNajmabadi 
> Just as I can use expressions to build up arbitrarily complex patterns (including collections), I want the same ability to break apart arbitrary graphs (including collections) with patterns.

Are you sure that this is what end users need, rather than something that you C# team developers need?

**Comment by mikernet** [comment]
Pattern matching, especially when combined with collection pattern matching, is incredibly powerful and succinct. It can easily turn 20 lines of difficult to follow code into a simple switch expression with a few lines. I use it all the time in normal app development.

**Comment by mikernet** [comment]
There are lots of examples in the search Cyrus provided above. It's not hard to skim over the ones that aren't relevant and find good examples in those search results.

**Comment by fandrei** [comment]
@mikernet 

> It's not hard to skim over the ones that aren't relevant and find good examples

It's also easy to find some pretty awful examples.

> It can easily turn 20 lines of difficult to follow code into a simple switch expression with a few lines.

I'd love to see an example, but it looks like we digressed too much. Probably better to move this discussion to somewhere else.

**Comment by mikernet** [comment]
Does it even need to be implemented? You can achieve similar results with DIMs on interfaces. Where is the lack of this feature overly burdensome?

```cs
interface IBase
{
    object GetValue();
}

interface ISub : IBase
{
    new string GetValue();
    object IBase.GetValue() => GetValue();
}

class C : ISub
{
    public string GetValue() => "Wooo";
}
```

**Comment by mikernet** [comment]
It was a huge pain with classes because you couldn't override a base abstract method and provide a more specific one with the same name. It's not an issue in interfaces.

**Comment by fandrei** [comment]
So, a few examples of people asking for a solution of this "unimportant" problem:

https://github.com/dotnet/csharplang/discussions/4372
https://github.com/dotnet/csharplang/issues/2844
https://github.com/dotnet/csharplang/discussions/296
https://github.com/dotnet/csharplang/issues/2589
https://github.com/dotnet/csharplang/issues/930

**Comment by fandrei** [comment]
I think that simply auto-adding stub methods would be enough. Unlikely that changing runtime is really needed.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi 
> > Just as I can use expressions to build up arbitrarily complex patterns (including collections), I want the same ability to break apart arbitrary graphs (including collections) with patterns.
> 
> Are you sure that this is what end users need, rather than something that you C# team developers need?

Nope. I'm not sure. But what of it? Ultimately, certainty is not the goal.

**Comment by CyrusNajmabadi** [comment]
> There are lots of examples in the search Cyrus provided above. 

That was Jared. But we are often easily mistaken.

**Comment by CyrusNajmabadi** [comment]
> I think that simply auto-adding stub methods would be enough. Unlikely that changing runtime is really needed.

This is possible today with source generators. If this is enough, if advise you to just implement that as you have a solution today (which could work across all version I if c#), versus waiting years for us to do it.

**Comment by CyrusNajmabadi** [comment]
> I think that simply auto-adding stub methods would be enough. Unlikely that changing runtime is really needed.

This is possible today with source generators. If this is enough, if advise you to just implement that as you have a solution today (which could work across all version I if c#), versus waiting years for us to do it.

**Comment by mikernet** [comment]
@Charlieface I don't see it as hacky, it's relatively simple. I don't really see the problem, but that's just me. Even if `override` was provided, you would still be able to work around the expected results:

```
interface IBase
{
    object GetValue();
}

interface ISub : IBase
{
    override string GetValue();
}

class Evil : ISub, IBase
{
    public string GetValue() => "Foo";
    object IBase.GetValue() => "Bar";
}
```

**Comment by fandrei** [comment]
@CyrusNajmabadi gosh. I wasn't looking for _any_solution, there is a number of obvious ones. I just want this thing to be fixed finally.

**Comment by HaloFour** [comment]
@mikernet 

> It was a huge pain with classes because you couldn't override a base abstract method and provide a more specific one with the same name. It's not an issue in interfaces.

If you really want to blow minds about the whole "consistency" argument, .NET has always allowed for "explicit" overrides, C# has just never supported them.  Covariant returns might have not been necessary at all if it did.  😁

**Comment by CyrusNajmabadi** [comment]
> @jaredpar it also matches a lot of the code that constructs strings, and the code that could be easily rewritten in a different way,

Why does that matter? Lots of code can be rewritten in a different way. Indeed, that applies to most code. 

> I wanted to ask if you have a way to know how much different features are actually used, using a real C# parser, but it looks like you don't.

If that's what you want to know, then ask that. 

> Anyway, I don't remember many people asking about list pattern matching in C# on StackOverflow

I don't see how that's relevant. We don't design asking ourselves how much a particular person remembered feedback on StackOverflow.  We design based on what we think will provide the most value to the ecosystem as a whole. This will always be subjective, and it's entirely true that for every single release there will be people that disagree with the decisions made. This - applies to us and literally ever mainstream language ever created. 

I'm not sure what you're hoping to gain here. We thought investing in patterns was worthwhile.  Nothing has changed our opinions here. Similarly, investing in covariant returns for interface implementation has not been worthwhile. 

These are our choices for the past, and relitigating them won't change them. Am you can do is try to work with us on future decisions.  You're interested in certain things. The best of option for you is to make the best cases for investment there, and see if the community and designers agree with you.

**Comment by CyrusNajmabadi** [comment]
> along with its XML doc. 

Inheritdoc is a thing. 

> which requires duplicating every such function

Here's the thing. You consider that not acceptable. But we may disagree. Ultimately, we have to choose which tiny handful of features we pick. And if we think the workarounds are not so bad, then c'est la vie. 

Language design and scenario prioritization age ultimately entirely subjective processes. Your subjective assessment differs from ours. That's life. 

--

Reminder, my subjective opinion differs from many on the team. I likely have more rejected proposals than nearly anyone else out there. That's not something that bothers me. It just means figuring out better how to convince people and make good specifications for the ecosystem as a whole.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi gosh. I wasn't looking for _any_solution, there is a number of obvious ones. I just want this thing to be fixed finally.

Then find a way for this to be viewed as not important than all the other things that we and the ecosystem want "fixed finally" :-)

I'm ok with the status quo. If you're not, then I wish you best of luck on what to do about that.

**Comment by CyrusNajmabadi** [comment]
> I think inheritdoc won't work with a new function, you'd still need to copy it once for each duplicated function.

You can just place it wherever you want, including on new functions.

**Comment by fandrei** [comment]
@CyrusNajmabadi 
> Then find a way for this to be viewed as not important than all the other things that we and the ecosystem want "fixed finally" :-)

You probably wanted to say "more important". I already listed some of the issues created for this problem before.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi 
> > Then find a way for this to be viewed as not important than all the other things that we and the ecosystem want "fixed finally" :-)
> 
> You probably wanted to say "more important". I already listed some of the issues created for this problem before.

Phone autocorrect strikes again

**Comment by CyrusNajmabadi** [comment]
> already listed some of the issues created for this problem before.

Great. We'll consider this when evaluating our backlog.

**Comment by CyrusNajmabadi** [comment]
> > Great. We'll consider this when evaluating our backlog.
> 
> How often does that evaluation happen? I struggled to get any clear info on it re a different issue.
> Like is there an approx cycle to it, or once per langver etc?
> 
> 

It's fairly continuous. Occasionally we take a little break to do nothing but reevaluation.

**Comment by mikernet** [comment]
> No, that would be impossible, because they are the same "slot". CLR uses a slot-style vtable, so an `override` takes over the slot of the base member, which cannot be further overridden separately from the derived's member.

I highly doubt that is how this would work. If you look at my code, you'll see that the class states that it implements both interfaces, so it can specify how it implements the base interface member:

```cs
class Evil : ISub, **IBase**
```

It's kind of similar (but not exactly the same) as how you can do:

```cs
public class MyList<T> : List<T>, ICollection
{
    int Count => -999;
}
```

I doubt that this feature would be designed such that you would be blocked from redeclaring an interface implementation.

**Comment by mikernet** [comment]
> I think inheritdoc won't work with a `new` function, you'd still need to copy it once for each duplicated function.

inheritdoc has a `cref` attribute that can point at anything you want - a method in an entirely different class, a base implementation, etc. It doesn't need to point at something that it is actually inheriting.

**Comment by mikernet** [comment]
Inheritdoc doesn't even need to match signatures exactly. For example, it can point to a method with more parameters and it'll still match the common ones.

Some fun examples:

https://github.com/Singulink/Singulink.Enums/blob/13e79195166f4178bce6432305aaa1c224451eca/Source/Singulink.Enums/EnumConverter.cs#L119

https://github.com/Singulink/Singulink.Numerics.BigDecimal/blob/0f53af521db9b135ec21e10d0f823de5e37fbd87/Source/Singulink.Numerics.BigDecimal/BigDecimal.Parsing.cs
